pkg/v1/remote/referrers_test.go

```diff
@@ -0,0 +1,185 @@
+			ArtifactType: "application/testing123",
+		}
+	}
+
+	// Push an image we'll attach things to.
+	// We'll copy from src to dst.
+	rootRef, err := name.ParseReference(fmt.Sprintf("%s/repo:root", u.Host))
+	if err != nil {
+		t.Fatal(err)
+	}
+	rootImg, err := random.Image(10, 10)
+	if err != nil {
+		t.Fatal(err)
+	}
+	rootImg = mutate.ConfigMediaType(rootImg, types.MediaType("application/testing123"))
+	if err := remote.Write(rootRef, rootImg); err != nil {
+		t.Fatal(err)
+	}
+	rootDesc := descriptor(rootImg)
+	t.Logf("root image is %s", rootDesc.Digest)
+
+	// Push an image that refers to the root image as its subject.
+	leafRef, err := name.ParseReference(fmt.Sprintf("%s/repo:leaf", u.Host))
+	if err != nil {
+		t.Fatal(err)
+	}
+	leafImg, err := random.Image(20, 20)
+	if err != nil {
+		t.Fatal(err)
+	}
+	leafImg = mutate.ConfigMediaType(leafImg, types.MediaType("application/testing123"))
+	leafImg = mutate.Subject(leafImg, rootDesc).(v1.Image)
+	if err := remote.Write(leafRef, leafImg); err != nil {
+		t.Fatal(err)
+	}
+	leafDesc := descriptor(leafImg)
+	t.Logf("leaf image is %s", leafDesc.Digest)
+
+	// Get the referrers of the root image, by digest.
+	rootRefDigest := rootRef.Context().Digest(rootDesc.Digest.String())
+	index, err := remote.Referrers(rootRefDigest)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if d := cmp.Diff([]v1.Descriptor{leafDesc}, index.Manifests); d != "" {
+		t.Fatalf("referrers diff (-want,+got): %s", d)
+	}
+
+	// Get the referrers by querying the root image's fallback tag directly.
+	tag, err := name.ParseReference(fmt.Sprintf("%s/repo:sha256-%s", u.Host, rootDesc.Digest.Hex))
+	if err != nil {
+		t.Fatal(err)
+	}
+	idx, err := remote.Index(tag)
+	if err != nil {
+		t.Fatal(err)
+	}
+	mf, err := idx.IndexManifest()
+	if err != nil {
+		t.Fatal(err)
+	}
+	if d := cmp.Diff(index.Manifests, mf.Manifests); d != "" {
+		t.Fatalf("fallback tag diff (-want,+got): %s", d)
+	}
+
+	// Push the leaf image again, this time with a different tag.
+	// This shouldn't add another item to the root image's referrers,
+	// because it's the same digest.
+	// Push an image that refers to the root image as its subject.
+	leaf2Ref, err := name.ParseReference(fmt.Sprintf("%s/repo:leaf2", u.Host))
+	if err != nil {
+		t.Fatal(err)
+	}
+	if err := remote.Write(leaf2Ref, leafImg); err != nil {
+		t.Fatal(err)
+	}
+	// Get the referrers of the root image again, which should only have one entry.
+	rootRefDigest = rootRef.Context().Digest(rootDesc.Digest.String())
+	index, err = remote.Referrers(rootRefDigest)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if d := cmp.Diff([]v1.Descriptor{leafDesc}, index.Manifests); d != "" {
+		t.Fatalf("referrers diff after second push (-want,+got): %s", d)
+	}
+
+	// Try applying filters and verify number of manifests and and annotations
+	goodFilter := map[string]string{types.OCIFilterArtifactType: "application/testing123"}
+	index, err = remote.Referrers(rootRefDigest, remote.WithFilter(goodFilter))
```

Comment:
Can we change the method signature of `WithFilter` to something like `WithFilter(key, value)` and allow it to be passed potentially multiple times to build up the same map?

```
remote.Referrers(d,
    remote.WithFilter("type", "sbom"),
    remote.WithFilter("flavor", "salty"))
```

Reply:
yee